include/deal.II/dofs/dof_tools.h

```diff
@@ -1368,20 +1368,44 @@ namespace DoFTools
    * This function is used in step-31, step-32, and step-42, for example.
    */
   template <int dim, int spacedim>
-  void
+  std::vector<std::vector<bool>>
   extract_constant_modes(const DoFHandler<dim, spacedim> &dof_handler,
-                         const ComponentMask             &component_mask,
-                         std::vector<std::vector<bool>>  &constant_modes);
+                         const ComponentMask             &component_mask);
+
+  /**
+   * Same as above.
+   *
+   * @deprecated
+   */
+  template <int dim, int spacedim>
+  DEAL_II_DEPRECATED_WITH_COMMENT(
+    "Use the other function that return the constant modes.")
+  void extract_constant_modes(const DoFHandler<dim, spacedim> &dof_handler,
+                              const ComponentMask             &component_mask,
+                              std::vector<std::vector<bool>>  &constant_modes);
 
   /**
    * Same as above but for multigrid levels.
    */
   template <int dim, int spacedim>
-  void
+  std::vector<std::vector<bool>>
   extract_level_constant_modes(const unsigned int               level,
                                const DoFHandler<dim, spacedim> &dof_handler,
-                               const ComponentMask             &component_mask,
-                               std::vector<std::vector<bool>>  &constant_modes);
+                               const ComponentMask             &component_mask);
```

Comment:
Have you considered whether returning an `IndexSet` (or vector of `IndexSet`s) instead might be a better way?

Reply:
Using `IndexSet` will likely (unless we sort DoFs per component) be considerably more memory-intensive than a vector of bools. Not that it matters much in the grand scheme. The more important comment I have with respect to the interface is the analogy to #17477 which requires vectors of the same size, and how the vectors are "consumed" inside the multigrid functions. So I am not sure if a change in interface as suggested here is worth it. I would personally keep the vector of bit vectors, but I do not feel strongly about it.

Reply:
> The more important comment I have with respect to the interface is the analogy to https://github.com/dealii/dealii/pull/17477 which requires vectors of the same size, and how the vectors are "consumed" inside the multigrid functions.

I agree with this and also would keep the vectors of vectors.

Reply:
OK.

Reply:
I should add that I was primarily thinking of this in the context of parallel programs where we have to do some contortions for `std::vector<bool>` (we need to translate between global and local indices) which we would not have to do with an `IndexSet`. But I'm good with keeping things as is.